loader/src/main.py

```diff
@@ -172,11 +174,26 @@ def process_event(msg):
             # references OR is referenced by other documents)
             logger.debug(
                 f"Resolving references {analysis.reference_paths}",
-                extra={"resource_id": resource_id},
+                extra={"batch_id": batch_id, "resource_id": resource_id},
+            )
+
+            # Redis distributed lock for reference binding
+            dlm = Redlock(
+                [{
+                    "host": REDIS_REFERENCES_HOST,
+                    "port": REDIS_REFERENCES_PORT,
+                    "db": REDIS_REFERENCES_DB
+                }, ]
             )
+            while True:
+                # Lock TTL is 5 min
+                ref_lock = dlm.lock(f"batch:{batch_id}:lock", 3600_000)
+                if ref_lock:
+                    break
             resolved_fhir_instance = binder.resolve_references(
```

Comment:
What if a exception is raised in here and the lock is never unlocked? 🙁

Reply:
That's why the lock has a ttl

Reply:
Oh right!
Isn't there a way to ensure that the unlock will be called? Something like a `defer` but I'm not sure Python has that